### DISCUSSION TITLE
Problems when trying to run FlexRic and xApp with a COTS UE

### DISCUSSION THREAD

**Discussion by jorgebaranda** [srsran/srsRAN_Project#567] [General Help]
Hi SRSers,
First of all, let me send a lot of kudos to the developers and the community for this amazing work. 
I am opening this discussion to report my findings and seek suggestions about the following issue.
I was able to succesfully follow this [tutorial](https://docs.srsran.com/projects/project/en/latest/tutorials/source/flexric/source/index.html)  and connect an srsUE to report metrics to the mentioned xApp (although most of the time the reported value is 0).
My problem comes when I am trying to move this setup to an over-the-air (OTA) setup running the gNB using the UHD driver to interact with an USRPB210 and using a COTS UE (Realme 9 smartphone). 
Following the tutorial steps, I see how the gNB registers to the FlexRIC. Then, I attach the Realme9 to the mobile network. The UE has connectivity (I can use the browser app to visit websites or I can use the Youtube app to see videos).  
However, when I launch the xApp, I see how the FlexRIC crashes with the following error: 
nearRT-RIC: /FlexRIC/flexric/src/sm/kpm_sm/kpm_sm_v02.03/dec/dec_asn_kpm_common/dec_meas_data.c:20: kpm_dec_meas_data_asn: Assertion `meas_data[i].meas_record_len >= 1 && meas_data[i].meas_record_len <= maxnoofMeasurementValue' failed.

Has someone else tried to use a COTS UE with srsran(UHD)+Open5gs core + Flexric? Is it compatible srs gNB code "using UHD, OFH" with FlexRIC or any other RIC (like the one by OSC)? Has the kind of UE influence in the setup? 

Again, thank you very much for your help and guidance and above all for such impressive work.
Best regards,
Jordi

**Comment by pgawlowicz** [comment]
Hi Jordi, 
could you share your gNB config and e2 pcap file? 
BR, 
Piotr

**Reply by jorgebaranda** [reply]
Hi Piotr,
Here you have the gNB config file and two pcap captures reproducing the issue in a compressed folder. The setup and the steps are the following ones.

- Setup: In the computer with IP address 10.1.14.54, we run the gnB srsRAN (commit bcf941b34), the flexric (branch br-flexric, commit ddb0a6a) and the xApp using the config file you included in your website tutorial.

- Steps:  First, I start the FlexRiC. Second, I start the gNB and I see how it contacts the FlexRIC. Third, I attach the Realme9 smarthpone and I check there is connectivity by watching a youtube video. Fourth, I start the xApp and I see crashing the FlexRic program with the error reported in the first post.  

Best regards,
Jordi Baranda
[srsRAN_FlexRIC_COTSUE.zip](https://github.com/srsran/srsRAN_Project/files/15064043/srsRAN_FlexRIC_COTSUE.zip)

**Reply by pgawlowicz** [reply]
Hi Jordi,

Could you try with the newest srsRAN-Project relese (commit: [1483bda](https://github.com/srsran/srsRAN_Project/commit/1483bda3091420cf7270eacdf31de932865c6294)) and the newest Flexric `br-flexric` branch?

Please compile flexric with `cmake -DKPM_VERSION=KPM_V3 -DXAPP_DB=NONE_XAPP ../`

In addition, you will need to apply this small patch to our repo, as it seems that FlexRIC does not like `granul_period` to be present in a RIC indication message.

```
diff --git a/lib/e2/e2sm/e2sm_kpm/e2sm_kpm_report_service_impl.cpp b/lib/e2/e2sm/e2sm_kpm/e2sm_kpm_report_service_impl.cpp
index f56757380..466b287c3 100644
--- a/lib/e2/e2sm/e2sm_kpm/e2sm_kpm_report_service_impl.cpp
+++ b/lib/e2/e2sm/e2sm_kpm/e2sm_kpm_report_service_impl.cpp
@@ -144,8 +144,8 @@ void e2sm_kpm_report_service_style1::clear_collect_measurements()
 bool e2sm_kpm_report_service_style1::collect_measurements()
 {
   // Set the granularity period.
-  ric_ind_message.granul_period_present = true;
-  ric_ind_message.granul_period         = granul_period;
+  ric_ind_message.granul_period_present = false;
+  //ric_ind_message.granul_period         = granul_period;
 
   // Fill indication msg.
   std::vector<meas_record_item_c> meas_records_items;
```

Let me know if it helps. 
BR,
Piotr

**Reply by jorgebaranda** [reply]
Hi Piotr,
Thanks for your answer. Regarding your suggestions, I have a light question.
Reviewing the code versions, I am using the srsRAN-Project commit  (commit bcf941b) available in the master repo when the "FlexRIC tutorial" was updated back in February. The Flexric code of br-flexric branch I am using  is in the last commit (no need to update it). 
Checking in the current srsRAN code (commit bcf941b), the change you mention in file (e2sm_kpm_report_service_impl.cpp), it is already like this (so this has been changed in the last srs code release). 

Then, if I am not wrong, since code versions that I have should be valid for this setup (srs+flexric), would it be a matter of compiling the Flexric code with the flag -DKPM_VERSION=KPM_V3 (previously, I did it with flag KPM_v2 as mentioned in previous tutorial release) and test again? Or is there any other enhancement on the E2 interface done in the last srsRAN release that should be considered and can affect?

The problem I face is when I use a Realme9 smartphone as a UE, that the FlexRIC crashed when starting the xApp. When, I do the test with a "ZMQ user" the whole setup worked, although the reported values in the xAPP were always 0 (maybe because of low traffic using ping).     

Thanks again for your help and guidance,
BR,
Jordi

**Reply by pgawlowicz** [reply]
Hi Jordi,

No, the old code supports only KPM_V2, so it will not work with the newest Flexric compiled with `-DKPM_VERSION=KPM_V3`.

In the newest version of our code, we always include the `granul_period` in the `RIC Indication Message`, while in the old version that you use the  `granul_period` parameter was not present to support Flexric directly. Now, we decided to always include it and contacted Flexric developers to fix the issue. 

I would recommend using the newest versions of both projects and apply the above patch to srsRAN-Project. It will be needed until Flexric fixes the issue. 

Best,
Piotr

**Comment by jorgebaranda** [comment]
Hi Piotr,
I did the test you mentioned with the commits of the srsRAN project and FlexRIC, the patch and compiling FlexRIC using KPM_V3. I experienced the same behaviour I explained in my first post: with ZMQ the xApp is reporting 0 values (sending a ping to 8.8.8.8 using the srsue UE) and with the Realme9 UE using a b210 usrp and watching a youtube video, the FlexRiC program crashes with the same error ("now in version kpm3") when I start the xApp :
"nearRT-RIC: /home/oran/flexric/src/sm/kpm_sm/kpm_sm_v03.00/dec/dec_asn_kpm_common/dec_meas_data.c:20: kpm_dec_meas_data_asn: Assertion `meas_data[i].meas_record_len >= 1 && meas_data[i].meas_record_len <= maxnoofMeasurementValue' failed.
Aborted (core dumped)" 

Attached you can find all the configurations files for the gnb (using ursp), flexric and xApp. For the case of flexric and xApp, I did the test using the loopback interface (127.0.0.1) and the IP of the NIC, but with identical result as you can see in the attached traces. 

The kind of UE shouldn't affect, isn't it? I can not understand why this difference when using srsue (zmq) and a smartphone UE with usrp. 

Best regards,
Jordi Baranda
[srsRAN_FlexRIC_traces_updatedcode.zip](https://github.com/srsran/srsRAN_Project/files/15151473/srsRAN_FlexRIC_traces_updatedcode.zip)

**Comment by pgawlowicz** [comment]
Hi Jordi,
Could you add the following block to your gnb config and try again?

```
metrics:
  rlc_json_enable: 1                # Enable RLC metrics reporting
  rlc_report_period: 1000           # Set reporting period to 1s
```

Could you also capture the e2ap packets exchanged between flexric and gnb?

Best,
Piotr

**Reply by jorgebaranda** [reply]
Dear Piotr,
Thanks thanks a lot for spotting this issue. Indeed, this was the key point of everything. :) After inserting these lines in the configuration file of the gnb, everything worked both for kpm_v2 and kpm_v3. It is worth mentioning about the code snippet above about the granul_period. If using Flexric, this has to be set to "false" until Flexric team solve it. 
Let me paste here the final configuration file of the gnb (in .txt format) for other people in the future with the same problem. 
Again thank you very much and best regards,
[gnb_final_config_file.txt](https://github.com/srsran/srsRAN_Project/files/15159774/gnb_final_config_file.txt)

Jordi Baranda

**Reply by pgawlowicz** [reply]
Nice! thanks for your feedback!
